### ISSUE TITLE
ico icon not working

### ISSUE THREAD

**Issue by Vzz1c** [end2endzone/ShellAnything#150]
I managed to simulate the full behavior of the `Beyond Compare` contetxt menu using the `<visibility>` tag! It's amazing! But I have a problem, using ico icon address directly doesn't work, but setting `dll+index` does, does it have anything to do with the icon size? I've tried different sizes and it doesn't work, please provide the correct example, thanks!

**Environment**
win10 22H2 Screen ratio 125%

**Comment by end2endzone** [comment]
Hi. I apologize for the late reply.

The feature for assigning an icon to a menu should be pretty strait forward. The documentation is available in the [UserManual, icons section](https://github.com/end2endzone/ShellAnything/blob/master/UserManual.md#icons).

> using ico icon address directly doesn't work

So for example, you should only have to specify the following (without an index) :
```xml
<icon path="C:\Program Files\ShellAnything\docs\shellanything.ico" />
```

> does it have anything to do with the icon size?

Maybe. I am not sure. I noticed you specified that you had scaling enabled at 125% for your screen. I recently had an issue with scaling that I described in #125. However I do not think this is what you are experiencing. In #125, the icons were always visible but scaled incorrectly. You said that you tried multiple sizes and none worked.

I also know that icons 16x16 up to 256x256 have been tested. I think icons beyond 256x256 resolution are stored as png and not bitmap, so there might be issues with resolutions above 256x256. For testing on your side, you can try [the icons from the tests directory](https://github.com/end2endzone/ShellAnything/tree/master/src/tests/test_files). These are confirmed working fine.

Is you icon multi-layered? I do not know the exact term but for example, with Gimp, you can save each layers individually in a single ico file. I do not think these are handled properly by ShellAnything.

For helping with the investigation...

1. Could you confirm that you are using the latest release version or a locally build version?
2. Could you post a copy of the actual icon here? I could try it also from my side.
3. Have you [looked at the logs](https://github.com/end2endzone/ShellAnything/blob/master/UserManual.md#logging-support)? If there is an issue when reading the icon, you should get a _warning_ or an _error_, you should see files such as `sa.shellextension.dll.WARNING.20191125-083701.9456.log` or `sa.shellextension.dll.ERROR.20191125-084152.9456.log`. Take a look at them and see if there are issues when loading your icon. Logs can contain sensitive data, do not post a full log here 😉
4. The code that loads an icon is partially located [in CContextMenu.cpp#L175](https://github.com/end2endzone/ShellAnything/blob/0.9.0/src/shellextension/CContextMenu.cpp#L175). You may notice that the code at line 186 calls Win32Utils::GetIconSize() which is [implemented here](https://github.com/end2endzone/ShellAnything/blob/0.9.0/src/shared/Win32Utils.cpp#L394). The "normal" icon size should be 16x16 (multiplied by your scaling). For you particular scenario, the "best icon size" would be 20x20. However, Windows should pick the "most appropriate size" for your need.
5. The icon is converted as a bitmap at line 186 ([implemented here](https://github.com/end2endzone/ShellAnything/blob/0.9.0/src/shared/Win32Utils.cpp#L606)). Maybe there is an error in the conversion. If you can supply the actual icon, maybe I can validate how the code is behaving with this particular icon.

Waiting for your feedback.

**Comment by Vzz1c** [comment]
I found the reason, `resource_hacker` software extracted ioc icon, any size can not be used, I change the icon extraction software can be, but there is a point, use, use `ioc` icon, must be followed by the `index="0"`, otherwise it will not take effect, even if the use of `test_files` directory `ioc` icon is also the same. ioc` icon in the `test_files` directory.

**Comment by end2endzone** [comment]
I am sorry, I am having difficulties to understand what you are saying. Also, you keep saying `ioc` instead of `ico`. I assume this is multiple typos.

I think what you are saying is that there is a bug with icon specification. With the latest release, an icon element **_must have_** `index="0"` in the xml document to be displayed on a menu.

According to [the documentation about icons](https://github.com/end2endzone/ShellAnything/blob/master/UserManual.md#icons) :

    "The index is optional for ico file format."
    "If the index is not specified, the value 0 is used."

In other words, all the following should be valid icon definitions :
```xml
<icon path="c:\windows\system32\ieframe.dll" />
<icon path="C:\Users\JohnDoe\Documents\foobar.ico" />
```

This is definitely a bug! Nice catch!

**Comment by end2endzone** [comment]
I investigated this issue and found out the following :
* The [code that parses icons](https://github.com/end2endzone/ShellAnything/blob/0.9.0/src/core/ObjectFactory.cpp#L620) do not set the default **_Index 0_** when index attribute is not specified. The code leave icons without an implicit index to -1.
* The [code that validates icons](https://github.com/end2endzone/ShellAnything/blob/0.9.0/src/core/Icon.cpp#L66) says icons with _**Index -1**_ are invalid.
